docs: object detection example (#280)
aica-technology/api: docs/docs/examples/guides/object_detection_example.md

---
sidebar_position: 6
title: Using YOLO to Track Objects
---

import rvizgif from './assets/object-detection-example-rviz.gif'

# Using YOLO to track objects

This example provides a use case for `Object Detection Components`. We show how to create a custom component which
converts bounding boxes into a 3D pose. The example works with either an Intel Realsense camera using
`collections/intel-realsense-collection` or the Camera Streamer from `components/core-vision`. The final application will be able to track an object with a robot based on a fixed camera position.

<div class="text--center">

>   <img src={rvizgif} alt="Moving the robot towards an object in RVis" />

Reviewer: ```suggestion
  <img src={rvizgif} alt="Moving the robot towards an object in RViz" />
```

</div>

> 

Reviewer: ```suggestion
```

## Setup
To run a YOLO executor, you need a yolo model file in ONNX format and a YAML class file. Yolo models are widely available in PT formats. For example, download the lightweight YOLO12n.pt from ultralytics [here](https://github.com/sunsmarterjie/yolov12).

To convert a PT file to ONNX, run the following Python code with ultralytics installed:

```python
from ultralytics import YOLO
model = YOLO("yolov12n.pt")
model.export(format="onnx")  # creates 'yolov12n.onnx'
```

For this example, also download the standard coco.yaml class file [here](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/cfg/datasets/coco.yaml).
Move the ONNX model file and the YAML class file into a new directory.

In AICA Launcher, create a configuration with the following core version and packages:

> - AICA Core v4.3.2

Reviewer: ```suggestion
- AICA Core v4.4.2
```

might as well

- `collections/object-detection` (TODO: specify version)
- `collections/intel-realsense-collection v1.0.0`
- `components/core-vision v1.0.0`

Under Advanced Settings, add a volume containing the folder where you stored your YOLO model files and link it to `/files`.

## Using the YOLO Executor

The *YOLO Executor* component observes runs the YOLO segmentation model on an image. 
It takes a camera stream and outputs the segmented image as well as the locations of bounding boxes on the image. The user can set the `Object Class` parameter, which will cause the component to set the predicate `Object Detected` to `True` when the specified object is found in the image. The `Confidence Threshold` parameter is the minimum score a predicted bounding box must have to be considered a valid detection. `IOU Threshold` is used during Non-Maximum Suppression (NMS) to decide whether two bounding boxes represent the same object. For example, if `IOU threshold` is set to 0.5, any box that overlaps more than 50% with a higher-scoring box will be discarded. For now we leave them as default values.

To test the YOLO executor:
- Remove the `*Hardware Interface*`.
- Add the following components:

>   - `Realsense2 Camera/Realsense Camera` if you have an Intel Realsense camera.

Reviewer: I think we can Just add the component name (not the category name) to remove a bit of visual clutter, but I'm not sure everyone will agree. For example, we could perhaps mention the components and their category title in the setup, and from that point forward just mention them by component name only

    Otherwise, use the `Core Vision Components/Camera Streamer`, setting the `source` parameter.
    - Turn on the **auto-configure** and **auto-activate** switches
  - `Object Detection Components/YOLO Executor`
    - Set `Model Path` to the `.onnx` file, e.g., `/files/yolo12n.onnx`
    - Set `Classes Path` to the yaml label file, e.g., `/files/coco.yaml`
    - Set `Rate` to 3 (works on most machines), on a machine with a GPU this can be set to higher.
    - Turn on the **auto-configure** and **auto-activate** switches
- Connect both components’ _load nodes_ to load on program start
- Connect the `Image` output of the `Realsense2 Camera`/`Camera Streamer` to the `RGB Image` input of the *YOLO Executor*

The complete application is shown below:
![Graph](./assets/object-detection-using-yolo-graph.png)

The YOLO Executors parameters are as follows:
![Graph](./assets/object-detection-yolo-parameters.png)

> 

Reviewer: ```suggestion
```

### Running the application
Start the application in AICA Studio. Then open **RViz**: bottom-right gear icon → "Launch RViz", then RViz → Add → By
topic → `/yolo_executor/annotated_image/Image` to view the YOLO model's annotated output. It should show the camera
images with bounding boxes drawn around key objects in it. The bounding boxes are published on the
`yolo_executor/bounding_boxes` topic as `std_msgs/msg/String` and can be viewed in AICA Studio in the "ROS Topics" tab.
Changing the `Confidence Threshold` and `IOU Threshold` parameters will change how often and how many bounding boxes the
model outputs.

:::note

> Only users with a Linux host can visualize the robot with RViz. On macOS, AICA Launcher will not show the RViz option.

Reviewer: ```suggestion
Only users with a Linux host can visualize the image stream RViz. On macOS, AICA Launcher will not show the RViz option.
```

:::

## Tracking an object with YOLO

The bounding boxes generated by YOLO can be used to move a robot towards an object.

### Creating a custom component to estimate position from bounding boxes

We first need to create a custom component which can estimate a 3D pose from a position in the camera frame. More
information about custom components can be found [here](https://docs.aica.tech/docs/category/custom-components/). We
make 3 simplifying assumptions:

1. Camera is fixed and angled downward
2. Known camera position pointing down: `(x=0, y=0.6, z=0.6)`
3. Objects are on a flat surface at `z=0`

#### Set up the repository

- Create a git repository from the [component-template](https://github.com/aica-technology/component-template)
- Clone the repository, enter the directory, and run:

  ```bash
  ./initialize_package.sh
  ```

> 

Reviewer: ```suggestion
```

`ROS Topics/tf_to_signal/pose(modulo_interfaces/msg/EncodedState)`.

#### Transforming a frame to world

To drive the robot towards the object position it needs to be in `world` frame. The *Cartesian Transformation* component can convert from `camera_frame` to `world`. Add it to the application and connect the load node. Finally, connect the `pose` output from *TF to Signal* to `Input 1`, and the output of *YOLO to Marker* to `Input 2`. The output of *Cartesian Transformation* is now the object position in the world frame.

#### Moving a robot towards an object

We can use a *Signal Point Attractor* to move the robot end effector towards the object. Add the component to the
application, and set it to **auto-configure** and **auto-activate**. The load node can be connected to the `On Activate`
transition in *YOLO to Marker*. This means that the robot will not move until after *YOLO to Marker* has started. The

> `Cartesian State`, under *Robot State Broadcaster* should be connected to the `Input State`_` of *Signal Point Attractor*,

Reviewer: ```suggestion
`Cartesian State`, under *Robot State Broadcaster* should be connected to the `Input State` of *Signal Point Attractor*,
```

            transition: activate
        on_load:
          lifecycle:
            component: signal_point_attractor
            transition: configure
    parameters:
      rate: !!float 2.0
    inputs:
      state: /hardware/robot_state_broadcaster/cartesian_state
      attractor: /cartesian_transformation/output
    outputs:
      twist: /signal_point_attractor/twist
  cartesian_transformation:
    component: aica_core_components::signal::CartesianTransformation
    display_name: Cartesian Transformation
    inputs:
      input_1: /tf_to_signal/pose
      input_2: /yolo_to_marker/state_command
    outputs:
      output: /cartesian_transformation/output
  tf_to_signal:
    component: aica_core_components::ros::TfToSignal
    display_name: TF to Signal
    events:
      transitions:
        on_configure:
          lifecycle:
            component: tf_to_signal
            transition: activate
        on_load:
          lifecycle:
            component: tf_to_signal
            transition: configure
          load:
            component: cartesian_transformation
    parameters:
      frame: camera_frame
    outputs:
      pose: /tf_to_signal/pose
  video_player_component:

>     component: component_utils::VideoPlayer

Reviewer: this is no longer valid and should be replaced with the Camera Streamer references. Technically that also warrants for a new screenshot of the app

---
sidebar_position: 6

> title

Reviewer 2: This frontmatter is invalid and will cause Docusaurus to fail

```suggestion
title: Using YOLO to Track Objects
```

---

import rvizgif from './assets/object-detection-example-rviz.gif'

# Using YOLO to Track Objects

This example provides a use case for `Object Detection Components`. We show how to create a custom component which
converts bounding boxes into a 3D pose. The example works with either an Intel Realsense camera using
`collections/intel-realsense-collection` or a video played through a custom component which is provided. The final application will be able to track an object with a robot based on a fixed camera position.

<div class="text--center">
  <img src={rvizgif} alt="Moving the robot towards an object in RVis" />
</div>

## Setup

In AICA Launcher, create a configuration with the following core version and packages:

- AICA Core v4.3.2
- `collections/object-detection` (TODO: specify version)

> - `collections/intel-realsense-collection v1.0.0`

Reviewer 2: TODO: use the generic camera streamer from core-vision components once they are released (@bpapaspyros)

- Set `Rate` to 3 (works on most machines)

The parameter `Confidence Threshold` is the minimum score a predicted bounding box must have to be considered a valid detection. `IOU Threshold` is used during Non-Maximum Suppression (NMS) to decide whether two bounding boxes represent the same object. For example, if `IOU threshold` is set to 0.5, any box that overlaps more than 50% with a higher-scoring box will be discarded. For now we leave them as default values.

:::info

**Converting YOLO Model files to ONNX format**

The *YOLO Executor* component expects an ONNX file. These can be created from a pretrained [
`.pt`](https://github.com/sunsmarterjie/yolov12) model (we use the lightweight version — YOLO12n).

Run the following Python script (with [`ultralytics`](https://pypi.org/project/ultralytics/) installed) to convert a
`.pt` model to `.onnx`:

```python
from ultralytics import YOLO
model = YOLO("yolov12n.pt")
model.export(format="onnx")  # creates 'yolov12n.onnx'
```

:::

These YOLO models are pre-trained on coco - classes file for
found [here](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/cfg/datasets/coco.yaml). A fine-tuned

> model may require a custom yaml file.

Reviewer 2: When testing, I found this a bit hard to follow. There are enough links and information here on the page to get everything done but it's not obvious which links I need to follow. In particular, it was not obvious to me that `[``.pt``](https://github.com/sunsmarterjie/yolov12)` was the main link for the yolov12 models. Also, if I already was configuring AICA Launcher and then realized I needed to mount files, I needed to jump around the sections.

I would put this information in a section near the top, for example "## Preparing YOLO models", and structure it in a super simple way for the minimum default demo:
"To run a YOLO detector, you need a yolo model file in ONNX format and a YAML class file. Yolo models are widely available in PT formats. For example, download the lightweight YOLO12n.pt from ultralytics [here]. 

To convert a PT file to ONNX, run the following Python code with ultralytics installed:
...

For this example, also download the standard coco.yaml class file [here].

Move the ONNX model file and the YAML class file into a new directory.
"

This then makes it easier for the user to prepare their files before starting the AICA Launcher mount configuration.

After going through the minimum example, I would then follow up later on the page to briefly explain what fine-tuning a model with different class file or datasets would look like, but that it ultimately would be the same process for the component in the end.
